src/pubsub/mod_pubsub_db_rdbms.erl

```diff
@@ -0,0 +1,609 @@
+                      ItemId :: mod_pubsub:itemId()) -> iolist().
+sql_insert_item(Nidx, LU, LS, ItemId) ->
+    ["INSERT INTO pubsub_items (nidx, luser, lserver, itemid)"
+    " VALUES (", esc_int(Nidx), ", ",
+                 esc_string(LU), ", ",
+                 esc_string(LS), ", ",
+                 esc_string(ItemId), ")"].
+
+-spec sql_get_entity_items(Nidx :: mod_pubsub:nodeIdx(),
+                           LU :: jid:luser(),
+                           LS :: jid:lserver()) -> iolist().
+sql_get_entity_items(Nidx, LU, LS) ->
+    ["SELECT itemid"
+    " FROM pubsub_items"
+    " WHERE nidx = ", esc_int(Nidx),
+      " AND luser = ", esc_string(LU),
+      " AND lserver = ", esc_string(LS) ].
+
+-spec sql_delete_item(Nidx :: mod_pubsub:nodeIdx(),
+                      LU :: jid:luser(),
+                      LS :: jid:lserver(),
+                      ItemId :: mod_pubsub:itemId()) -> iolist().
+sql_delete_item(Nidx, LU, LS, ItemId) ->
+    ["DELETE FROM pubsub_items"
+    " WHERE nidx = ", esc_int(Nidx),
+      " AND luser = ", esc_string(LU),
+      " AND lserver = ", esc_string(LS),
+      " AND itemid = ", esc_string(ItemId) ].
+
+-spec sql_delete_all_items(Nidx :: mod_pubsub:nodeIdx()) -> iolist().
+sql_delete_all_items(Nidx) ->
+    ["DELETE FROM pubsub_items"
+    " WHERE nidx = ", esc_int(Nidx)].
+
+%%====================================================================
+%% Helpers
+%%====================================================================
+
+-spec delete_affiliation_wo_subs_check(Nidx :: mod_pubsub:nodeIdx(),
+                                       LU :: jid:luser(),
+                                       LS :: jid:lserver()) -> ok.
+delete_affiliation_wo_subs_check(Nidx, LU, LS) ->
+    SQL = sql_delete_affiliation(Nidx, LU, LS),
+    {updated, _} = mongoose_rdbms:sql_query(global, SQL),
+    ok.
+
+-type item_row() :: { Nidx :: mod_pubsub:nodeIdx(),
+                      LU :: binary(),
+                      LS :: binary(),
+                      ItemId :: binary() }.
+-type aff_row() :: { Nidx :: mod_pubsub:nodeIdx(),
+                     LU :: binary(),
+                     LS :: binary(),
+                     AffInt :: integer() }.
+-type sub_row() :: { Nidx :: mod_pubsub:nodeIdx(),
+                     LU :: binary(),
+                     LS :: binary(),
+                     LR :: binary(),
+                     TypeInt :: integer(),
+                     SubId :: binary() }.
+
+-spec build_states(ItemRows :: [item_row()], AffRows :: [aff_row()], SubRows :: [sub_row()]) ->
+    [mod_pubsub:pubsubState()].
+build_states(ItemRows, AffRows, SubRows) ->
+    Result1 = item_rows_to_states(ItemRows, #{}),
+    Result2 = aff_rows_to_states(AffRows, Result1),
+    maps:values(sub_rows_to_states(SubRows, Result2)).
+
+item_rows_to_states([], Acc) ->
+    Acc;
+item_rows_to_states([{ Nidx, LU, LS, ItemId } | RRows], Acc) ->
+    LJID = { LU, LS, <<>> },
+    #pubsub_state{ items = Items0 }
```

Comment:
This  and following line are not very readable (at least to me). I think the following would be a bit better:
```suggestion
    PS = maps:get({LJID, Nidx}, Acc, #pubsub_state{ stateid = {LJID, Nidx} }),
    #pubsub_state{ items = Items0 } = PS,
```